### DISCUSSION TITLE
How to run the build version in dev machine?

### DISCUSSION THREAD

**Discussion by dncpax** [lowdefy/lowdefy#722] [Questions and Help] [answered]
Hi again.

So my question is really about understanding the under-pinnings of lowdefy. 
In dev we have the capability to run an app with npx lowdefy dev.
So there must be also the capability to run in production, right? Like running the build version?

Thanks.

**Comment by Gervwyk** [comment]
Hi @dncpax

Awesome let’s get into a little more detail. Lowdefy runs a node express server which basically serves the page config and the app assets.

The build step for Lowdefy only prepares the app into some json config files. You can view these files in your .lowdefy folder when running the dev server. 

The app server is stateless which means it store no data or cache between requests. This allows you to easily start and stop servers as requires (docker), even run apps in serverless environments like aws lambda or netlify functions. For deployment details, checkout the Deployments tab in the menu on https://docs.lowdefy.com

So why do we need a server? Primarily for all reasons auth related (user auth, role based access control, etc.) and to parse requests secrets. This allows us to connect securely to databases, APIs and so on. It's also a great place to execute proprietary logic since only request ids are served to the client and no other request details. Also users can only triggers requests in Lowdefy which exists on pages to which they have access. 

So because the app server only serves json blobs to the client and all app logic is executed in the @lowdefy/engine package, wouldn't it be great if Lowdefy apps could also be static pages with no server? Yes. This would be great and super useful in many cases like just creating web pages, wikis docs, etc. 

We plan to support this soon, we just need to do a few additional things in terms of server setup and page load time optimization to get there 🚀

Did you have a specific use case in mind for a Lowdefy app?

**Comment by SamTolmay** [comment, marked as answer]
Hi @dncpax 

Maybe I can elaborate a little on what @Gervwyk said.

As he mentioned, the Lowdefy build process converts the configuration you have written into JSON configuration files that the Lowdefy server can use. These build artefacts can then be served by any Lowdefy server.

Currently, we provide 4 different Lowdefy servers. 

The development server is part of the CLI. This is the server that runs when you run `npx lowdefy dev`. This server is not optimised for production, and it watches the file system of your project repository, and rebuilds and reloads the web client if it detects a change.

We provide 3 production servers, A Docker server that runs as a docker image, and then servers that are set up to run in the AWS Lambda and Netlify serverless environments. You can read more on how to deploy them at [https://docs.lowdefy.com](https://docs.lowdefy.com).

What we don't currently have, and this is maybe what you are looking for, is a way to run a production Lowdefy server as a Node.js process. I think this is something we should do, and it shouldn't be too hard. Please let me know if this is the case?

**Reply by dncpax** [reply]
Yes, I am looking at ways of running lowdefy apps.
It seems to me that if the CLI can run apps, then it would seem it could be somewhat mimicked to a streamlined version, eg that doesn't monitor for changes for instance, or runs as a process with PM2...
I'm happy to wait until something like this comes along.

Thanks to both for great answers.

**Reply by SamTolmay** [reply]
This is exactly what we were thinking. I have opened an issue so you can track this request: #723.

**Comment by SamTolmay** [comment]
We released a production Node.js server in version 3.20. You can read more [here](https://docs.lowdefy.com/node-server).

We have not yet set up an example repository, but I will put one up soon.
